src/EditorFeatures/Core/Shared/Extensions/IRefactorNotifyServiceExtensions.cs

```diff
@@ -2,48 +2,109 @@
             }
-
-            return true;
         }
 
-        public static bool TryOnAfterGlobalSymbolRenamed(
-            this IEnumerable<IRefactorNotifyService> refactorNotifyServices,
+        private static async Task TryNotifyChangesAsync(
+            this IEnumerable<Lazy<IRefactorNotifyService>> refactorNotifyServices,
             Workspace workspace,
-            IEnumerable<DocumentId> changedDocuments,
-            ISymbol symbol,
-            string newName,
-            bool throwOnFailure)
+            Solution newSolution,
+            Solution oldSolution,
+            CancellationToken cancellationToken)
         {
-            foreach (var refactorNotifyService in refactorNotifyServices)
+            var projectChanges = newSolution.GetChanges(oldSolution).GetProjectChanges().ToImmutableArray();
+            var changedDocumentIds = projectChanges.SelectMany(pd => pd.GetChangedDocuments(onlyGetDocumentsWithTextChanges: true)).ToImmutableArray();
+            var _ = PooledDictionary<ISymbol, ISymbol>.GetInstance(out var changedSymbols);
+
+            foreach (var documentId in changedDocumentIds)
             {
-                if (!refactorNotifyService.TryOnAfterGlobalSymbolRenamed(workspace, changedDocuments, symbol, newName, throwOnFailure))
+                var newDocument = newSolution.GetRequiredDocument(documentId);
+                var oldDocument = oldSolution.GetDocument(documentId);
+
+                if (oldDocument is null)
                 {
-                    return false;
+                    continue;
+                }
+
+                var newSemanticModel = await newDocument.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var oldSemanticModel = await oldDocument.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+                var renamedNodes = await GatherAnnotatedNodesAsync(newDocument, cancellationToken).ConfigureAwait(false);
+
+                foreach (var node in renamedNodes)
+                {
+                    foreach (var annotation in node.GetAnnotations(RenameSymbolAnnotation.RenameSymbolKind))
+                    {
+                        var oldSymbol = annotation.ResolveSymbol(oldSemanticModel.Compilation);
+                        Contract.ThrowIfNull(oldSymbol);
+
+                        var newSymbol = newSemanticModel.GetDeclaredSymbol(node);
+                        Contract.ThrowIfNull(newSymbol);
+
+                        changedSymbols.Add(oldSymbol, newSymbol);
+                    }
                 }
             }
 
-            return true;
+            foreach (var (oldSymbol, newSymbol) in changedSymbols)
+            {
+                foreach (var refactorNotifyService in refactorNotifyServices)
+                {
+                    if (refactorNotifyService.Value.TryOnBeforeGlobalSymbolRenamed(workspace, changedDocumentIds, oldSymbol, newSymbol.Name, false))
```

Comment:
So we're no longer worrying about raising the "before" before any actual edits happen?